modules/core/src/main/java/org/apache/ignite/internal/table/criteria/SqlSerializer.java

```diff
@@ -0,0 +1,221 @@
+ * the License. You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.table.criteria;
+
+import static org.apache.ignite.internal.util.StringUtils.nullOrBlank;
+
+import java.util.Collection;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import org.apache.ignite.internal.util.CollectionUtils;
+import org.apache.ignite.table.criteria.Column;
+import org.apache.ignite.table.criteria.Criteria;
+import org.apache.ignite.table.criteria.CriteriaVisitor;
+import org.apache.ignite.table.criteria.Expression;
+import org.apache.ignite.table.criteria.Operator;
+import org.apache.ignite.table.criteria.Parameter;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Serializes {@link Criteria} into into SQL.
+ *
+ * <p>Note: Doesn't required any context to traverse an criteria tree, {@code null} can be used as initial context.
+ */
+public class SqlSerializer implements CriteriaVisitor<Void> {
+    private static final Map<Operator, String> ELEMENT_TEMPLATES = Map.of(
+            Operator.EQ, "{0} = {1}",
+            Operator.IS_NULL, "{0} IS NULL",
+            Operator.IS_NOT_NULL, "{0} IS NOT NULL",
+            Operator.GOE, "{0} >= {1}",
+            Operator.GT, "{0} > {1}",
+            Operator.LOE, "{0} <= {1}",
+            Operator.LT, "{0} < {1}",
+            Operator.NOT, "NOT {0}"
+    );
+
+    @SuppressWarnings("StringBufferField")
+    private final StringBuilder builder = new StringBuilder(128);
+
+    private final List<Object> arguments = new LinkedList<>();
+
+    /**
+     * Get query arguments.
+     *
+     * @return Query arguments.
+     */
+    public Object[] getArguments() {
+        return arguments.toArray(new Object[0]);
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public <T> void visit(Parameter<T> argument, @Nullable Void context) {
+        append("?");
+
+        arguments.add(argument.getValue());
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public <T> void visit(Column column, @Nullable Void context) {
+        append(column.getName());
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public <T> void visit(Expression expression, @Nullable Void context) {
+        Operator operator = expression.getOperator();
+        Criteria[] elements = expression.getElements();
+
+        String template;
+
+        if (operator == Operator.AND || operator == Operator.OR) {
```

Comment:
we shoud have more tests for the code. As I see not covered cases with one parameter, for three parameters, no test for OR, and so on....

Reply:
Extended